server/src/test/java/org/elasticsearch/cluster/routing/IndexRoutingTableTests.java

```diff
@@ -0,0 +1,183 @@
+        );
+        shardTable2 = new IndexShardRoutingTable(
+            p2,
+            List.of(getShard(p2, true, ShardRoutingState.UNASSIGNED, ShardRouting.Role.INDEX_ONLY))
+        );
+        indexRoutingTable = new IndexRoutingTable(index, new IndexShardRoutingTable[] { shardTable1, shardTable2 });
+        assertFalse(indexRoutingTable.readyForSearch(clusterState));
+
+        // 2 primaries that are index only with replicas that are not all available
+        shardTable1 = new IndexShardRoutingTable(
+            p1,
+            List.of(
+                getShard(p1, true, ShardRoutingState.STARTED, ShardRouting.Role.INDEX_ONLY),
+                getShard(p1, false, ShardRoutingState.STARTED, ShardRouting.Role.SEARCH_ONLY),
+                getShard(p1, false, ShardRoutingState.STARTED, ShardRouting.Role.SEARCH_ONLY)
+            )
+        );
+        shardTable2 = new IndexShardRoutingTable(
+            p2,
+            List.of(
+                getShard(p2, true, ShardRoutingState.STARTED, ShardRouting.Role.INDEX_ONLY),
+                getShard(p2, false, ShardRoutingState.UNASSIGNED, ShardRouting.Role.SEARCH_ONLY),
+                getShard(p2, false, ShardRoutingState.UNASSIGNED, ShardRouting.Role.SEARCH_ONLY)
+            )
+        );
+        indexRoutingTable = new IndexRoutingTable(index, new IndexShardRoutingTable[] { shardTable1, shardTable2 });
+        if (fastRefresh) {
+            assertTrue(indexRoutingTable.readyForSearch(clusterState));
+        } else {
+            assertFalse(indexRoutingTable.readyForSearch(clusterState));
+        }
+
+        // 2 primaries that are index only with some replicas that are all available
+        shardTable1 = new IndexShardRoutingTable(
+            p1,
+            List.of(
+                getShard(p1, true, ShardRoutingState.STARTED, ShardRouting.Role.INDEX_ONLY),
+                getShard(p1, false, ShardRoutingState.STARTED, ShardRouting.Role.SEARCH_ONLY),
+                getShard(p1, false, ShardRoutingState.STARTED, ShardRouting.Role.SEARCH_ONLY)
+            )
+        );
+        shardTable2 = new IndexShardRoutingTable(
+            p2,
+            List.of(
+                getShard(p2, true, ShardRoutingState.STARTED, ShardRouting.Role.INDEX_ONLY),
+                getShard(p2, false, ShardRoutingState.STARTED, ShardRouting.Role.SEARCH_ONLY),
+                getShard(p2, false, ShardRoutingState.STARTED, ShardRouting.Role.SEARCH_ONLY)
+            )
+        );
+        indexRoutingTable = new IndexRoutingTable(index, new IndexShardRoutingTable[] { shardTable1, shardTable2 });
+        assertTrue(indexRoutingTable.readyForSearch(clusterState));
+
+        // 2 unassigned primaries that are index only with some replicas that are all available
+        // TODO: Is this correct behavior for fastRefresh ? Fast refresh indices do not support replicas so this can not practically happen.
```

Comment:
nit This is the correct behavior currently for fastRefresh; search/replica shards do not do anything. Indeed, it may change behavior in the future, but I believe this test should then fail when we try to do that, and we'll simply change it. Thus I am not sure the TODO is needed.